src/components/Timer.react.js

```diff
@@ -0,0 +1,322 @@
+            prevProps.duration !== this.props.duration ||
+            prevProps.max_intervals !== this.props.max_intervals ||
+            prevProps.reset !== this.props.reset ||
+            prevProps.rerun !== this.props.rerun ||
+            prevProps.messages !== this.props.messages ||
+            prevProps.mode !== this.props.mode
+        ) {
+            this.resetTimer(this.props);
+        } else {
+            this.handleTimer(this.props);
+        }
+    }
+
+    componentWillUnmount() {
+        this.clearTimer();
+    }
+
+
+    render() {
+        const {id, class_name, style} = this.props;
+        return <div
+                id={id}
+                style={style}
+                className={class_name}
+            >
+                <div>{this.renderMessage}</div>
+            </div>
+
+    }
+}
+
+Timer.propTypes = {
+    /**
+     * The ID of this component, used to identify dash components
+     * in callbacks. The ID needs to be unique across all of the
+     * components in an app.
+     */
+    id: PropTypes.string,
+
+    /**
+     * This component will increment the counter `n_intervals` every
+     * `interval` milliseconds
+     */
+    interval: PropTypes.number,
+
+    /**
+     * If True, the n_interval counter  and the timer no longer updates.  This pauses the timer.
+     */
+    disabled: PropTypes.bool,
+
+    /**
+     * Number of times the interval has passed (read-only)
+     */
+    n_intervals: PropTypes.number,
+
+    /**
+     * Number of times the interval will be fired.
+     * If -1, then the interval has no limit (the default)
+     * and if 0 then the interval stops running.
+     */
+    max_intervals: PropTypes.number,
+
+    /**
+     * When in countdown mode, the timer will count down to zero from the starting `duration` and will show the number
+     *  of milliseconds remaining.
+     *  When in stopwatch mode, the timer will count up from zero and show the number of milliseconds elapsed.
+     *  (read-only)
+     */
+    timer: PropTypes.number,
+
+    /**
+     * The timer will count down to zero in `countdown` mode and count up from zero in `stopwatch` mode
+     */
+    mode: PropTypes.oneOf(['stopwatch', 'countdown']),
+
+    /**
+     * Sets the number of milliseconds the timer will run.  If -1 the timer will not be limited by the duration. (the default)
+     * and if 0 then the timer stops running.
+     */
+    duration: PropTypes.number,
+
+    /**
+     * This will start the timer at the beginning with the given prop settings.
+     */
+    reset: PropTypes.bool,
+
+    /**
+     * A list of the time(s) in milliseconds at which to fire a callback. This can be used to start a task at a given
+     * time rather than using the timer. Since the timer is typically set at a small interval like one second, using
+     * fire_times can reduce the number of times a callback is fired and can increase app performance. The time(s) must be a
+     * multiple of the interval.
```

Comment:
```suggestion
     * fire_times can reduce the number of times a callback is fired and can increase app performance. Each time must be a
     * multiple of the interval.
```
I wonder though if we can't relax that constraint - and update `at_fire_time` as soon as we've reached or passed the specified point. I can see this being useful for example if you're fiddling with `interval` to see how it affects the displayed values, it could be frustrating to need to adjust `fire_times` every time you do that - especially since a matching adjustment may need to happen inside your callback as well.

Reply:
I can see the advantages of doing it that way.  The main problem is with the countdown timer, which you'd expect to stop at zero.  How do you make it stop at zero if the `duration` is not a multiple of the `interval`?

Or, if you don't make it stop at zero, do you stop it short or run the timer to a negative number?   Either would look like a bug.

Reply:
Right, it's a good deal more complicated if we allow `fire_times` and `duration` to not be multiples of `interval` - speaking of which we should document what happens with `duration` in this case. Looks to me like right now we just go past zero and then stop? That's fine for now as long as we document it, but later we may want to do better: either start or end with a one-off `setTimeout` of the remainder time. For `fire_times` I also think eventually we should allow them to occur between intervals, and in fact this could also be handled by a `setTimeout` created during the last full interval before the fire time. But again, not necessary for now, we can add that later.